Improve ordering of headers with clang-format

ximinez opened the pull request:
- Remove all manual header groupings from source and header files

Follow-up to #5293.

<!--
This PR template helps you to write a good pull request description.
Please feel free to include additional useful information even beyond what is requested below.

If your branch is on a personal fork and has a name that allows it to
run CI build/test jobs (e.g. "ci/foo"), remember to rename it BEFORE
opening the PR.  This avoids unnecessary redundant test runs. Renaming
the branch after opening the PR will close the PR.
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch
-->

## High Level Overview of Change

<!--
Please include a summary of the changes.
This may be a direct input to the release notes.
If too broad, please consider splitting into multiple PRs.
If a relevant task or issue, please link it here.
-->

### Context of Change

<!--
Please include the context of a change.
If a bug fix, when was the bug introduced? What was the behavior?
If a new feature, why was this architecture chosen? What were the alternatives?
If a refactor, how is this better than the previous implementation?

If there is a spec or design document for this feature, please link it here.
-->

### Type of Change

<!--
Please check [x] relevant options, delete irrelevant ones.
-->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Performance (increase or change in throughput and/or latency)
- [ ] Tests (you added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation update
- [ ] Chore (no impact to binary, e.g. `.gitignore`, formatting, dropping support for older tooling)
- [ ] Release

### API Impact

<!--
Please check [x] relevant options, delete irrelevant ones.

* If there is any impact to the public API methods (HTTP / WebSocket), please update https://github.com/xrplf/rippled/blob/develop/API-CHANGELOG.md
  * Update API-CHANGELOG.md and add the change directly in this PR by pushing to your PR branch.
* libxrpl: See https://github.com/XRPLF/rippled/blob/develop/docs/build/depend.md
* Peer Protocol: See https://xrpl.org/peer-protocol.html
-->

- [ ] Public API: New feature (new methods and/or new fields)
- [ ] Public API: Breaking change (in general, breaking changes should only impact the next api_version)
- [ ] `libxrpl` change (any change that may affect `libxrpl` or dependents of `libxrpl`)
- [ ] Peer protocol change (must be backward compatible or bump the peer protocol version)

<!--
## Before / After
If relevant, use this section for an English description of the change at a technical level.
If this change affects an API, examples should be included here.

For performance-impacting changes, please provide these details:
1. Is this a new feature, bug fix, or improvement to existing functionality?
2. What behavior/functionality does the change impact?
3. In what processing can the impact be measured? Be as specific as possible - e.g. RPC client call, payment transaction that involves LOB, AMM, caching, DB operations, etc.
4. Does this change affect concurrent processing - e.g. does it involve acquiring locks, multi-threaded processing, or async processing?
-->

<!--
## Test Plan
If helpful, please describe the tests that you ran to verify your changes and provide instructions so that others can reproduce.
This section may not be needed if your change includes thoroughly commented unit tests.
-->

<!--
## Future Tasks
For future tasks related to PR.
-->

bthomee commented on src/xrpld/nodestore/detail/codec.h:
```diff
@@ -28,10 +28,8 @@
-
-#include <lz4.h>
 #include <nudb/detail/field.hpp>
-
+#include <lz4.h>
```
This one is curious. LZ4 is an external dependency, and the letter 'l' comes before 'n' in the alphabet. Why is LZ4 still sorted after NuDB?

vvysokikh1 replied:
it's sorted by reg expr so it's unable to deduct what is actually external.

vvysokikh1 replied:
I actually remembered that this PR: https://github.com/XRPLF/rippled/pull/5216/files

has changes to reg exprs as well, and it seems to me that they handle lz4.h include properly: https://github.com/XRPLF/rippled/pull/5216/files#diff-63b89f03ecf9641e0eeadcecfe8121b42f8a5ea3f47090b8ea0f59197804ed0e

ximinez replied:
The decision is largely arbitrary, decided by the `IncludeCategories` in `.clang-format`. The ordering is determined by whichever of these regexes matches first:
1. '^<(test)/' - Anything starting with "test"
2. '^<(xrpld)/' - Anything starting with "xrpld"
3. '^<(xrpl)/' - Anything starting with "xrpl"
4. '^<(boost)/' - Anything starting with "boost"
5. '^.*/' - Anything with a slash "/", basically anything else that includes a path
6. '^.*\.h' - Anything with ".h"
7. '.*' - Everything else.

We could easily switch 5 and 6, or make them equivalent, but it made more sense to me that something with a path is likely to be a downloaded dependency (via Conan), and anything without a path, but with ".h" may be from Conan, or from the OS/system, while anything with neither is more likely to be a built-in C++ library header.

ximinez commented:
Added `IncludeBlocks: Regroup` and updated the formatting. It decreased the number of affected files from 664 to 621. Instead of mostly removing one or two empty lines, it looks like it's now mostly adding one or two empty lines to the files that _didn't_ have groupings.

I'm not going to do anything further with this PR. Someone else can feel free to grab it and take ownership. Or if it's ok as-is, we can just merge it.

vvysokikh1 reviewed (approved):
ship it

bthomee commented:
> ship it

The only thing I'm wondering about is if we should treat `test/`, `xrpl/` and `xrpld/` as a single group since those are all internal source files, i.e.

```
- Regex:           '^<(test|xrpld|xrpl)/'
   Priority:        0
```

Otherwise than that, I'm also ok with shipping it. I can make that change in .clang-format if it would be useful.

vvysokikh1 commented:
> > ship it
> 
> The only thing I'm wondering about is if we should treat `test/`, `xrpl/` and `xrpld/` as a single group since those are all internal source files, i.e.
> 
> ```
> - Regex:           '^<(test|xrpld|xrpl)/'
>    Priority:        0
> ```
> 
> Otherwise than that, I'm also ok with shipping it. I can make that change in .clang-format if it would be useful.

I think splitting them makes perfect sense, since we want to do a better separation of libxrpl and xrpld. Tests being in separate group also makes perfect sense

ximinez commented:
> > The only thing I'm wondering about is if we should treat `test/`, `xrpl/` and `xrpld/` as a single group since those are all internal source files, i.e.
> 
> I think splitting them makes perfect sense, since we want to do a better separation of libxrpl and xrpld. Tests being in separate group also makes perfect sense

I'm not going to do this, but what if we actually separated them even more? Specifically, order the `libxrpl` folders by levelization priority. Just a thought. It's probably not worth it, since cmake already enforces that there are no inappropriate dependencies.
